app/assets/javascripts/3dbio_viewer/src/webapp/components/model-upload/ModelUpload.tsx

```diff
@@ -0,0 +1,92 @@
+    const { compositionRoot } = useAppContext();
+
+    const [jobTitle, setJobTitle] = useState<string>("");
+    const [error, setError] = useState<string>();
+
+    const structureFileRef = useRef<DropzoneRef>(null);
+    const annotationFileRef = useRef<DropzoneRef>(null);
+
+    const onSubmitHandler = useCallback(() => {
+        setError("");
+        const structureFile = getFile(structureFileRef);
+        if (structureFile) {
+            const uploadParams = {
+                jobTitle,
+                structureFile,
+                annotationsFile: getFile(annotationFileRef),
+            };
+            return compositionRoot
+                .uploadAtomicStructure(uploadParams)
+                .run(result => result, console.error);
+        } else {
+            setError(i18n.t("Error: No file selected. Please select a structure file."));
+            return _.noop;
+        }
+    }, [compositionRoot, jobTitle]);
+
+    return (
+        <Dialog open={true} onClose={onClose} maxWidth="xl" fullWidth>
+            <DialogTitle>
+                {title}
+                <IconButton onClick={onClose}>
+                    <Close />
+                </IconButton>
+            </DialogTitle>
+
+            <DialogContent>
+                {error && <h3>{error}</h3>}
+                <p>
+                    {i18n.t(
+                        "Models under study and not deposited to PDB yet can be analysed too. Annotations from similar entries based on BLAST sequence match will be displayed, but also customised annotations can be provided by the user. Job title (if provided) will be used to identify the model, otherwise the file name will be used."
+                    )}
+                </p>
+
+                <label htmlFor="jobTitle">
+                    <strong>{i18n.t("Job Title")}</strong>
+                </label>
+                <small>{i18n.t("Optional")}</small>
+                <input
+                    aria-label={i18n.t("Job Title")}
+                    value={jobTitle}
+                    placeholder={i18n.t("Job Title")}
+                    onChange={e => setJobTitle(e.target.value)}
+                    id="jobTitle"
+                    type="text"
+                    className="form-control"
+                />
+
+                <label className="fileFormat">
+                    {i18n.t("Structure file in")}
+                    <a href="http://www.wwpdb.org/documentation/file-format"> PDB </a>
+                    {i18n.t("or")}
+                    <a href="http://mmcif.wwpdb.org/"> mmCIF </a> {i18n.t("format")}
+                </label>
+                <Dropzone ref={structureFileRef} accept=".pdb,.cif"></Dropzone>
+
+                <label className="fileFormat">{i18n.t("Upload your annotations")}</label>
+                <Dropzone ref={annotationFileRef} accept={"application/json"}></Dropzone>
+
+                <button className="uploadSubmit" onClick={useCallbackEffect(onSubmitHandler)}>
```

Comment:
Hook call should be outside the jsx block, `const submit = useCallbackEffect(onSubmitHandler)`. It's cleaner but also necessary, hooks must be executed always in the same order, and the JSX tag could be inside a conditional.